zkevm-circuits/src/blobkzg_circuit/src/main.rs

```diff
@@ -0,0 +1,585 @@
+                        ctx.next();
+                        integer_chip.mul(&mut ctx, &assigned_xns[i], &multiplier)?
+                    };
+                    ctx.next();
+                    integer_chip.assert_equal(&mut ctx, &assigned_xns[i + 1], &xn_next)?;
+                }
+
+                // 5. ADD CONSTRAINTS FOR EACH x^(2^i) STEP
+                for i in 0..255 {
+                    let x2i_next =
+                        integer_chip.mul(&mut ctx, &assigned_x2is[i], &assigned_x2is[i])?;
+                    ctx.next();
+                    integer_chip.assert_equal(&mut ctx, &assigned_x2is[i + 1], &x2i_next)?;
+                }
+                Ok(())
+            },
+        )?;
+        Ok(())
+    }
+}
+
+/// Native Field (likely BN256)
+type N = halo2_proofs::halo2curves::bn256::Fr;
+
+/// Wrong Field (BLS12-381)
+type W = halo2_proofs::halo2curves::bls12_381::Scalar;
+
+/// ROOT OF UNITY WITH ORDER 4096: 0x564c_0a11_a0f7_04f4_fc3e_8acf_e0f8_245f_0ad1_347b_378f_bf96_e206_da11_a5d3_6306
+const ROOT_OF_UNITY: W = W::from_raw([
+    0xe206_da11_a5d3_6306,
+    0x0ad1_347b_378f_bf96,
+    0xfc3e_8acf_e0f8_245f,
+    0x564c_0a11_a0f7_04f4,
+]);
+const ROOT_OF_UNITY_ORDER: W = W::from_raw([0x1000, 0x0, 0x0, 0x0]);
+
+const NUMBER_OF_POINTS: usize = 4096;
+
+fn main() {
+    let datapoints: [W; NUMBER_OF_POINTS] = std::array::from_fn(|i| W::from(i as u64 + 1));
+    let input_values = InputValues {
+        datapoints,
+        x: W::from(5),
+        w: ROOT_OF_UNITY,
+        n: ROOT_OF_UNITY_ORDER,
+    };
+    let circuit = BlobKZGCircuit::<W, N>::new(input_values);
+    let prover = halo2_proofs::dev::MockProver::<N>::run(20, &circuit, vec![vec![]]).unwrap();
+    prover.assert_satisfied();
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    fn random_field_value() -> W {
+        use halo2_proofs::arithmetic::Field;
+        W::random(rand_core::OsRng)
+    }
+    use rand::Rng;
+
+    #[test]
+    fn try_default() {
+        let datapoints = std::array::from_fn(|i| W::from(i as u64 + 1));
+        let input_values = InputValues {
+            datapoints,
+            x: W::from(5),
+            w: ROOT_OF_UNITY,
+            n: ROOT_OF_UNITY_ORDER,
+        };
+        let circuit = BlobKZGCircuit::<W, N>::new(input_values);
+        let prover = halo2_proofs::dev::MockProver::<N>::run(20, &circuit, vec![vec![]]).unwrap();
+        prover.assert_satisfied();
+    }
+
+    #[test]
+    fn try_random_datapoints() {
+        let datapoints = std::array::from_fn(|_| random_field_value());
+        let input_values = InputValues {
+            datapoints,
+            x: W::from(5),
+            w: ROOT_OF_UNITY,
+            n: ROOT_OF_UNITY_ORDER,
+        };
+        let circuit = BlobKZGCircuit::<W, N>::new(input_values);
+        let prover = halo2_proofs::dev::MockProver::<N>::run(20, &circuit, vec![vec![]]).unwrap();
+        prover.assert_satisfied();
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
Instead of panic, it might be better to expect a specific error variant, as the panic might be happening for reasons different than the expected. I believe `MockProver` will return error variants pointing to the constraint that failed - so we can be sure it is the right one